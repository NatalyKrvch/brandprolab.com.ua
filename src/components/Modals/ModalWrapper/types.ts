export interface ModalWrapperProps {
  isOpen: boolean;
  onClose: () => void;
  children: React.ReactNode;
  className?: string;
}
