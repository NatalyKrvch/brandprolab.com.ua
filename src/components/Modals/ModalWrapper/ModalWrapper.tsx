'use client';

import { createPortal } from 'react-dom';
import { IoCloseOutline } from 'react-icons/io5';
import { twMerge } from 'tailwind-merge';

import {
  MODAL_CLOSE_BUTTON_TEST_ID,
  MODAL_CONTENT_TEST_ID,
  MODAL_WRAPPER_TEST_ID,
} from '@/lib/testIDs';
import { ROUNDED_CLASSES } from '@/styles/constants';

import { useModalWrapper } from './hooks/useModalWrapper';
import type { ModalWrapperProps } from './types';

const ModalWrapper = ({
  isOpen,
  onClose,
  children,
  className,
}: ModalWrapperProps) => {
  const { mounted } = useModalWrapper(isOpen, onClose);

  if (!mounted || !isOpen) return null;

  return createPortal(
    <div
      data-testid={MODAL_WRAPPER_TEST_ID}
      className="fixed inset-0 z-50 flex items-center justify-center bg-teal-lightOpacity06 p-4"
      onClick={onClose}
    >
      <div
        className={twMerge(
          'relative flex w-full max-w-lg flex-col items-end bg-white p-6 shadow-lg',
          ROUNDED_CLASSES,
          className,
        )}
        data-testid={MODAL_CONTENT_TEST_ID}
        onClick={e => e.stopPropagation()}
      >
        <button onClick={onClose} data-testid={MODAL_CLOSE_BUTTON_TEST_ID}>
          <IoCloseOutline className="transition-color-fast h-7 w-7 text-gray-dark hover:text-teal-dark tablet:h-8 tablet:w-8 desktop:h-10 desktop:w-10" />
        </button>
        {children}
      </div>
    </div>,
    document.getElementById('modal-root')!,
  );
};

export default ModalWrapper;
