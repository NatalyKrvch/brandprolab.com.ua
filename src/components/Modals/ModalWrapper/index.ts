export { default as ModalWrapper } from './ModalWrapper';
export * from './types';
