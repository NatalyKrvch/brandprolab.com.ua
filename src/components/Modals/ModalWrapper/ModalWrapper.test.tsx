import '@testing-library/jest-dom';

import { fireEvent, render, screen } from '@testing-library/react';

import {
  MODAL_CLOSE_BUTTON_TEST_ID,
  MODAL_CONTENT_TEST_ID,
  MODAL_WRAPPER_TEST_ID,
} from '@/lib/testIDs';

import ModalWrapper from './ModalWrapper';

describe('ModalWrapper', () => {
  const modalRoot = document.createElement('div');
  modalRoot.setAttribute('id', 'modal-root');
  document.body.appendChild(modalRoot);

  const onClose = jest.fn();
  const childrenText = 'Modal content goes here';

  const renderModal = () =>
    render(
      <ModalWrapper isOpen={true} onClose={onClose}>
        <p>{childrenText}</p>
      </ModalWrapper>,
    );

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders modal with content', () => {
    const { asFragment } = renderModal();

    expect(screen.getByTestId(MODAL_WRAPPER_TEST_ID)).toBeInTheDocument();
    expect(screen.getByTestId(MODAL_CONTENT_TEST_ID)).toBeInTheDocument();
    expect(screen.getByText(childrenText)).toBeInTheDocument();

    expect(asFragment()).toMatchSnapshot();
  });

  it('calls onClose when clicking on backdrop', () => {
    const { asFragment } = renderModal();

    fireEvent.click(screen.getByTestId(MODAL_WRAPPER_TEST_ID));
    expect(onClose).toHaveBeenCalledTimes(1);

    expect(asFragment()).toMatchSnapshot();
  });

  it('calls onClose when clicking on close button', () => {
    const { asFragment } = renderModal();

    fireEvent.click(screen.getByTestId(MODAL_CLOSE_BUTTON_TEST_ID));
    expect(onClose).toHaveBeenCalledTimes(1);

    expect(asFragment()).toMatchSnapshot();
  });

  it('does not call onClose when clicking inside modal content', () => {
    const { asFragment } = renderModal();

    fireEvent.click(screen.getByTestId(MODAL_CONTENT_TEST_ID));
    expect(onClose).not.toHaveBeenCalled();

    expect(asFragment()).toMatchSnapshot();
  });
});
