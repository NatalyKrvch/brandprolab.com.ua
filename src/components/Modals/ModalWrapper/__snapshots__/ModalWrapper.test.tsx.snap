// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalWrapper calls onClose when clicking on backdrop 1`] = `<DocumentFragment />`;

exports[`ModalWrapper calls onClose when clicking on close button 1`] = `<DocumentFragment />`;

exports[`ModalWrapper does not call onClose when clicking inside modal content 1`] = `<DocumentFragment />`;

exports[`ModalWrapper renders modal with content 1`] = `<DocumentFragment />`;
